File: android/src/main/kotlin/com/logicwind/appsonair_flutter_applink/AppsonairFlutterApplinkPlugin.kt
package com.logicwind.appsonair_flutter_applink
import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.util.Log
import androidx.annotation.NonNull
import com.appsonair.applink.interfaces.AppLinkListener
import com.appsonair.applink.services.AppLinkService
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONObject

/** AppsonairFlutterApplinkPlugin */
class AppsonairFlutterApplinkPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {

  private lateinit var channel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private lateinit var referralEventChannel: EventChannel
  private var eventSink: EventChannel.EventSink? = null
  private lateinit var attributionEventChannel: EventChannel
  private var referralEventSink: EventChannel.EventSink? = null
  private var attributionEventSink: EventChannel.EventSink? = null
  private var activity: Activity? = null

  // initialize() runs from onAttachedToActivity, which can fire before Dart's initState() has
  // subscribed to these EventChannels. A callback that arrives before onListen would otherwise be
  // dropped silently (eventSink?.success() on a null sink is a no-op) even though native logs show
  // it fired. Buffer the latest payload per channel and flush it once a listener attaches.
  private var pendingDeepLink: String? = null
  private var pendingReferralInfo: String? = null
  private var pendingAttributionInfo: String? = null

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "appsOnAirAppLink")
    channel.setMethodCallHandler(this)

    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "appLinkEventChanel")
    referralEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "appLinkReferralEventChanel")
    attributionEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "appLinkAttributionEventChanel")

    eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
      override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        pendingDeepLink?.let {
          events?.success(it)
          pendingDeepLink = null
        }
      }

      override fun onCancel(arguments: Any?) {
        eventSink = null
      }
    })
    referralEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
      override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        referralEventSink = events
        pendingReferralInfo?.let {
          events?.success(it)
          pendingReferralInfo = null
        }
      }

      override fun onCancel(arguments: Any?) {
        referralEventSink = null
      }
    })
    attributionEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
      override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        attributionEventSink = events
        pendingAttributionInfo?.let {
          events?.success(it)
          pendingAttributionInfo = null
        }
      }

      override fun onCancel(arguments: Any?) {
        attributionEventSink = null
      }
    })
  }

  private fun createAppLink( result: Result, call: MethodCall){
    val appLinkService = activity?.let { AppLinkService.getInstance(it.applicationContext) }
    val url: String = call.argument<String>("url") ?: ""
    val name: String = call.argument<String>("name") ?: ""
    val urlPrefix: String = call.argument<String>("urlPrefix") ?: ""
    val shortId: String? = call.argument<String>("shortId") ?: null
    val androidFallbackUrl: String? = call.argument<String>("androidFallbackUrl") ?: null
    val iosFallbackUrl: String? = call.argument<String>("iosFallbackUrl") ?: null
    val socialMeta: Map<String, Any>? = call.argument<Map<String, Any>>("socialMeta") ?: null
    val isOpenInBrowserAndroid: Boolean? = call.argument<Boolean>("isOpenInBrowserAndroid") ?: null
    val isOpenInAndroidApp: Boolean? = call.argument<Boolean>("isOpenInAndroidApp") ?: null
    val isOpenInBrowserApple: Boolean? = call.argument<Boolean>("isOpenInBrowserApple") ?: null
    val isOpenInIosApp: Boolean? = call.argument<Boolean>("isOpenInIosApp") ?: null
    val appsFlyer: Map<String, Any>? = call.argument<Map<String, Any>>("appsFlyer") ?: null
    val attributionTtl: Int? = call.argument<Int>("attributionTtl") ?: null

    CoroutineScope(Dispatchers.Main).launch {
      val data = appLinkService?.createAppLink(
          name = name,
          url = url,
          urlPrefix = urlPrefix,
          shortId = shortId,
          socialMeta = socialMeta,
          androidFallbackUrl = androidFallbackUrl,
          iosFallbackUrl = iosFallbackUrl,
          isOpenInAndroidApp = isOpenInAndroidApp,
          isOpenInBrowserAndroid = isOpenInBrowserAndroid,
          isOpenInBrowserApple = isOpenInBrowserApple,
          isOpenInIosApp = isOpenInIosApp,
          appsFlyer = appsFlyer,
          attributionTtl = attributionTtl
      )
      result.success(data.toString())
    }
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
        "create_app_link" -> {
          activity?.let { createAppLink(result,call) }
        }
        "get_referral_details" -> {
          val appLinkService = activity?.let { AppLinkService.getInstance(it.applicationContext) }
          val referral = appLinkService?.getReferralDetails()
          result.success(referral.toString())
        }
       "get_referral_info" -> {
        val appLinkService = activity?.let { AppLinkService.getInstance(it.applicationContext) }
        CoroutineScope(Dispatchers.Main).launch {
          val referral = appLinkService?.getReferralInfo()
          result.success(referral.toString())
         }
       }
      "get_attribution_info" -> {
        val appLinkService = activity?.let { AppLinkService.getInstance(it.applicationContext) }
        CoroutineScope(Dispatchers.Main).launch {
          val attributionInfo = appLinkService?.getAttributionInfo()
          result.success(attributionInfo.toString())
         }
       }
        else -> {
          result.notImplemented()
        }
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    eventSink = null
    referralEventSink = null
    attributionEventSink = null
  }

  // Handle activity attachment and detachment
  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
    initializeAppLinkService(activity?.intent)
    binding.addOnNewIntentListener { intent ->
      handleNewIntent(intent)
      true
    }
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivity() {
    activity = null
  }

  private val appLinkListener = object : AppLinkListener {
    override fun onDeepLinkProcessed(uri: Uri, result: JSONObject) {
      val mapData = mapOf(
        "uri" to uri.toString(),
        "result" to result,
      )
      val payload = JSONObject(mapData).toString()
      val sink = eventSink
      if (sink != null) {
        sink.success(payload)
      } else {
        pendingDeepLink = payload
      }
    }

    override fun onReferralLinkDetected(result: JSONObject) {
      val payload = result.toString()
      val sink = referralEventSink
      if (sink != null) {
        sink.success(payload)
      } else {
        pendingReferralInfo = payload
      }
    }

    override fun onAttributionListener(result: JSONObject) {
      val payload = result.toString()
      val sink = attributionEventSink
      if (sink != null) {
        sink.success(payload)
      } else {
        pendingAttributionInfo = payload
      }
    }

    override fun onDeepLinkError(uri: Uri?, error: String) {
      Log.e("DeepLinkListener", "Failed to process deep link: $uri")
      Log.e("DeepLinkListener", "Error: $error")
    }
  }

  // One-time setup: registers the listener and processes the launch intent (if any). Per the SDK,
  // initialize() is idempotent and only processes the deep link on its first call — a later intent
  // arriving while the app is already running must go through handleDeepLink() instead, not another
  // initialize() call, otherwise it is silently dropped.
  private fun initializeAppLinkService(intent: Intent?) {
    val activity = activity ?: return
    val appLinkService = AppLinkService.getInstance(activity)
    appLinkService.initialize(activity, intent ?: Intent(), appLinkListener)
  }

  // Handle a deep link delivered to an already-running activity.
  private fun handleNewIntent(intent: Intent?) {
    val activity = activity ?: return
    if (intent == null) return
    val appLinkService = AppLinkService.getInstance(activity)
    appLinkService.handleDeepLink(intent, activity.packageName)
  }
}

File: ios/Classes/AppsonairFlutterApplinkPlugin.swift
import Flutter
import UIKit
import AppsOnAir_AppLink

public class AppsonairFlutterApplinkPlugin: NSObject, FlutterPlugin {
    
    // Singleton instance
    public static let shared = AppsonairFlutterApplinkPlugin()
    
    private var channel: FlutterMethodChannel?
    
    fileprivate static var pendingDeepLink: String?
    fileprivate static var pendingReferralDeepLink: String?
    
    private override init() {
        super.init()
        // Initialize AppLinkService once
        initializeAppLinkService()
    }
    
    private func initializeAppLinkService() {
        AppLinkService.shared.initialize { latestUrl, result in
            let jsonResponse = ["uri": latestUrl?.absoluteString ?? "", "result": result]
            let jsonData = try? JSONSerialization.data(withJSONObject: jsonResponse, options: [])
            let jsonString = String(data: jsonData ?? Data(), encoding: .utf8)
            
            if let sink = AppLinkEventHandler.shared.eventSink {
                sink(jsonString)
            } else {
                AppsonairFlutterApplinkPlugin.pendingDeepLink = jsonString
            }
        } onReferralLinkDetected: { result in
            let jsonData = try? JSONSerialization.data(withJSONObject: result, options: [])
            let jsonString = String(data: jsonData ?? Data(), encoding: .utf8)
            if let sink = ReferralEventHandler.shared.eventSink {
                sink(jsonString)
            } else {
                AppsonairFlutterApplinkPlugin.pendingReferralDeepLink = jsonString
            }
        }
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = AppsonairFlutterApplinkPlugin.shared
        instance.channel = FlutterMethodChannel(name: "appsOnAirAppLink", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: instance.channel!)
        
        // Each event channel gets its own handler
        let appLinkEventChannel = FlutterEventChannel(name: "appLinkEventChanel", binaryMessenger: registrar.messenger())
        appLinkEventChannel.setStreamHandler(AppLinkEventHandler.shared)
        
        let referralEventChannel = FlutterEventChannel(name: "appLinkReferralEventChanel", binaryMessenger: registrar.messenger())
        referralEventChannel.setStreamHandler(ReferralEventHandler.shared)
    }
    
    private func createAppLink(result: @escaping FlutterResult, call: FlutterMethodCall) {
        guard let args = call.arguments as? [String: Any] else {
            result("Invalid arguments")
            return
        }
        
        let url = args["url"] as? String ?? ""
        let name = args["name"] as? String ?? ""
        let shortId = args["shortId"] as? String
        let urlPrefix = args["urlPrefix"] as? String ?? ""
        let androidFallbackUrl = args["androidFallbackUrl"] as? String
        let iosFallbackUrl = args["iosFallbackUrl"] as? String
        let socialMeta = args["socialMeta"] as? [String: Any]
        
        let isOpenInBrowserAndroid = args["isOpenInBrowserAndroid"] as? Bool
        let isOpenInAndroidApp = args["isOpenInAndroidApp"] as? Bool
        let isOpenInBrowserApple = args["isOpenInBrowserApple"] as? Bool
        let isOpenInIosApp = args["isOpenInIosApp"] as? Bool
        
        AppLinkService.shared.createAppLink(
            url: url,
            name: name,
            urlPrefix: urlPrefix,
            shortId: shortId,
            socialMeta: socialMeta,
            isOpenInBrowserApple: isOpenInBrowserApple,
            isOpenInIosApp: isOpenInIosApp,
            iosFallbackUrl: iosFallbackUrl,
            isOpenInAndroidApp: isOpenInAndroidApp,
            isOpenInBrowserAndroid: isOpenInBrowserAndroid,
            androidFallbackUrl: androidFallbackUrl
        ) { latestLink in
            if let data = try? JSONSerialization.data(withJSONObject: latestLink, options: []),
               let jsonString = String(data: data, encoding: .utf8) {
                result(jsonString)
            } else {
                result("Error converting response to JSON string")
            }
        }
    }
    
    private func getReferralDetails(result: @escaping FlutterResult, call: FlutterMethodCall) {
        AppLinkService.shared.getReferralDetails {linkDetails in
            if let data = try? JSONSerialization.data(withJSONObject: linkDetails, options: []),
               let jsonString = String(data: data, encoding: .utf8) {
                result(jsonString)
            } else {
                result("Error converting response to JSON string")
            }
        }
           
    }

    private func getReferralInfo(result: @escaping FlutterResult, call: FlutterMethodCall) {
        AppLinkService.shared.getReferralInfo { linkDetails in
            if let data = try? JSONSerialization.data(withJSONObject: linkDetails, options: []),
               let jsonString = String(data: data, encoding: .utf8) {
                result(jsonString)
            } else {
                result("Error converting response to JSON string")
            }
        }
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "create_app_link":
            createAppLink(result: result, call: call)
        case "get_referral_details":
            getReferralDetails(result: result, call: call)
        case "get_referral_info":
            getReferralInfo(result: result, call: call)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

// MARK: - Separate Stream Handlers

class AppLinkEventHandler: NSObject, FlutterStreamHandler {
    static let shared = AppLinkEventHandler()
    var eventSink: FlutterEventSink?
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        if let pendingLink = AppsonairFlutterApplinkPlugin.pendingDeepLink {
            events(pendingLink)
            AppsonairFlutterApplinkPlugin.pendingDeepLink = nil
        }
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

class ReferralEventHandler: NSObject, FlutterStreamHandler {
    static let shared = ReferralEventHandler()
    var eventSink: FlutterEventSink?
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        if let pendingLink = AppsonairFlutterApplinkPlugin.pendingReferralDeepLink {
            events(pendingLink)
            AppsonairFlutterApplinkPlugin.pendingReferralDeepLink = nil
        }
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}
